File: FactoryMethod.cpp
/*****
 * Factory Method
 * Define an interface for creating an object, but let subclasses decide which class to instantiate.
 * Factory Method lets a class defer instantiation to subclasses.
*****/

#include <iostream>
#include <memory>

using namespace std;


class Product { public: virtual void method() = 0; };

class ConcreteProduct : public Product { public: void method() override { cout << "Concrete Product" << endl; } };

class Factory { public: virtual unique_ptr<Product> create() = 0; };

class ConcreteFactory : public Factory { public: unique_ptr<Product> create() override { return make_unique<ConcreteProduct>(); } };


int main() {
    auto f = make_unique<ConcreteFactory>();
    auto product = f->create();

    product->method();

    return 0;
}

File: FactoryMethod.py
"""
Factory Method
Define an interface for creating an object, but let subclasses decide which class to instantiate.
Factory Method lets a class defer instantiation to subclasses.
"""

from abc import ABC, abstractmethod


class Product(ABC):
    @abstractmethod
    def method(self):
        pass

class ConcreteProduct(Product):
    def method(self):
        print("Concrete Product")

class Factory(ABC):
    @abstractmethod
    def create(self) -> Product:
        pass

class ConcreteFactory(Factory):
    def create(self):
        return ConcreteProduct()


if __name__ == "__main__":
    f = ConcreteFactory()
    product = f.create()

    product.method()
